### PR TITLE
enhancement(kubernetes_logs source): Replace regex with a simple parser for CRI logs

### PR DESCRIPTION
**Pull request by Ilmarii**
This PR replaces the regex with a simple parser for the CRI log string.

The asymptotic complexity of the new parser does not depend on the length of the log.

This solves the issue of the growing number of open files mentioned in #15590
Also in our deployment with an average log length of 15 KB, this patch cuts the CPU load in half.
vector-cpu

### CODE REVIEW

File: src/sources/kubernetes_logs/parser/cri.rs [hunk 1]
```diff
@@ -173,6 +129,67 @@ impl FunctionTransform for Cri {
...
+#[inline]
+const fn is_delimiter(c: &u8) -> bool {
+    *c == b' '
+}
+
+/// Parses a CRI log line.
+///
+/// Equivalent to regex: `(?-u)^(?P<timestamp>.*) (?P<stream>(stdout|stderr)) (?P<multiline_tag>(P|F)) (?P<message>.*)(?P<new_line_tag>\n?)$`
+#[inline]
+fn parse_log_line(line: &[u8]) -> Option<ParsedLog> {
```
Comment (spencergilbert) [inline src/sources/kubernetes_logs/parser/cri.rs:149]: I don't think we want to be `expect()`ing in this function, as we'd want it to still return a `None` to hit the `emit!(ParserMatchError { value: &s[..] });` arm.

**Review by spencergilbert — CHANGES_REQUESTED** [review]
Appreciate the contribution! I haven't been super happy with the state of the parsers here and I'm excited to see such an improvement.

I do have one reservation about the existing implementation, otherwise it looks good to me!

File: src/sources/kubernetes_logs/parser/cri.rs (hunk 1 above)
Comment (Ilmarii) [reply src/sources/kubernetes_logs/parser/cri.rs:149]: Thank you for review! I will fix this part

**Review by spencergilbert — APPROVED** [review]
Awesome, thank you!

**Review by neuronull — APPROVED** [review]
LGTM :+1: 

Nice performance boost!

**Comment by spencergilbert** [comment]
Thanks again for the contribution @Ilmarii!

**Pull request merged**